### ISSUE TITLE
mysql issues after upgrade

### ISSUE THREAD

**Issue by bjo81** [NetKnights-GmbH/privacyidea-ucs#4]
After upgrading to the latest version from the UCS appcenter, a login is not possible any more:

```
Authentication failed. (pymysql.err.InternalError) (1054, u"Unknown column 'radiusserver.timeout' in 'field list'") [SQL: u'SELECT radiusserver.id AS radiusserver_id, radiusserver.identifier AS radiusserver_identifier, radiusserver.server AS radiusserver_server, radiusserver.port AS radiusserver_port, radiusserver.secret AS radiusserver_secret, radiusserver.dictionary AS radiusserver_dictionary, radiusserver.description AS radiusserver_description, radiusserver.timeout AS radiusserver_timeout, radiusserver.retries AS radiusserver_retries \nFROM radiusserver']
```

**Comment by cornelinux** [comment]
Then the schema upgrade script did not work correctly.
This one:
https://github.com/NetKnights-GmbH/privacyidea-ucs/blob/master/debian/privacyidea-ucs.postinst#L78

**Comment by bjo81** [comment]
On the contrary, I get the following errors on my test instance:
```
+ pi-manage db stamp 4f32a4e1bf33 -d /opt/privacyidea/privacyidea-venv/lib/privacyidea/migrations

             _                    _______  _______
   ___  ____(_)  _____ _______ __/  _/ _ \/ __/ _ |
  / _ \/ __/ / |/ / _ `/ __/ // // // // / _// __ |
 / .__/_/ /_/|___/\_,_/\__/\_, /___/____/___/_/ |_|
/_/                       /___/
   
Running online
+ pi-manage db upgrade -d /opt/privacyidea/privacyidea-venv/lib/privacyidea/migrations/

             _                    _______  _______
   ___  ____(_)  _____ _______ __/  _/ _ \/ __/ _ |
  / _ \/ __/ / |/ / _ `/ __/ // // // // / _// __ |
 / .__/_/ /_/|___/\_,_/\__/\_, /___/____/___/_/ |_|
/_/                       /___/
   
Running online
Column already exists
(pymysql.err.InternalError) (1060, u"Duplicate column name 'Type'") [SQL: u'ALTER TABLE tokeninfo ADD COLUMN `Type` VARCHAR(100)']
(pymysql.err.InternalError) (1060, u"Duplicate column name 'condition'") [SQL: u'ALTER TABLE policy ADD COLUMN `condition` INTEGER NOT NULL']
(pymysql.err.InternalError) (1060, u"Duplicate column name 'priority'") [SQL: u'ALTER TABLE resolverrealm ADD COLUMN priority INTEGER']
(pymysql.err.InternalError) (1060, u"Duplicate column name 'adminrealm'") [SQL: u'ALTER TABLE policy ADD COLUMN adminrealm VARCHAR(256)']
(pymysql.err.InternalError) (1060, u"Duplicate column name 'revoked'") [SQL: u'ALTER TABLE token ADD COLUMN revoked BOOL']
Could not add column 'locked' to table 'token'
(pymysql.err.InternalError) (1060, u"Duplicate column name 'locked'") [SQL: u'ALTER TABLE token ADD COLUMN locked BOOL']
(pymysql.err.InternalError) (1050, u"Table 'smtpserver' already exists") [SQL: u'\nCREATE TABLE smtpserver (\n\tid INTEGER NOT NULL AUTO_INCREMENT, \n\tidentifier VARCHAR(255) NOT NULL, \n\tserver VARCHAR(255) NOT NULL, \n\tport INTEGER, \n\tusername VARCHAR(255), \n\tpassword VARCHAR(255), \n\tsender VARCHAR(255), \n\
```
Where does `pi-manage db upgrade` get it's correct version to upgrade from?

**Comment by cornelinux** [comment]
The current version is stored in the database table alembic_version.
But, on older installations, this table does not exist. This is why we recommended to stamp the version to "4f32a4e1bf33" which is the first version of privacyIDEA 2.
All updates work this way, that they will do no harm, if applied several times. This is why we can stamp to the first version.

You can get the current version of you database like this:

      pi-manage db current -d /usr/lib/privacyidea/migrations/or/which/ever/path/

Get a history of the schema:

      pi-manage db current -d /usr/lib/privacyidea/migrations/or/which/ever/path/

or upgrade with ``upgrade``. However - in theory - in a correctly set up system you could skip the ``stamp`` and just run the ``upgrade``.

**Comment by bjo81** [comment]
Thanks for the explanation.
For a correctly setup system, I've put the app into a local test appcenter, and get now the following issue:
```
sqlalchemy.exc.InternalError: (pymysql.err.InternalError) (1071, u'Specified key was too long; max key length is 767 bytes') [SQL: u'\nCREATE TABLE smsgateway (\n\tid INTEGER NOT NULL AUTO_INCREMENT, \n\tidentifier VARCHAR(255) NOT NULL, \n\tdescription VARCHAR(1024), \n\tprovidermodule VARCHAR(1024) NOT NULL, \n\tPRIMARY KEY (id), \n\tUNIQUE (identifier)\n)\n\n']
```

**Comment by cornelinux** [comment]
Welcome to mariadb 10.1. :-/
We had this issue before...
https://github.com/privacyidea/privacyidea/issues/967
We have set the DYNAMIC in version 2.23, so this should^TM^ work out fine.
I will also investigate.

**Comment by cornelinux** [comment]
Hi @bjo81 

this solves the problem:

~~~~
# cat /etc/mysql/mariadb.conf.d/60-privacyidea.cnf 
[mysqld]
innodb_large_prefix=on 
innodb_file_format=Barracuda 
~~~~

**Comment by bjo81** [comment]
Great! UCS 4.3-2 still has mariadb 10.1.26, so this has to be set definitively.

**Comment by cornelinux** [comment]
Hah, I am curious, what will happen with updates. :-/
By the way, I added it to the test appcenter...

**Comment by cornelinux** [comment]
In my testing environment I successfully updated a UCS 4.2 with 2.21.4 to UCS 4.3 (Test Appcenter) with PI 2.23.dev8. ...using this current git state.

So in roughly one week we will release 2.23.1 and it will be available for UCS 4.3.
